pkg/runtime/types.go

```diff
@@ -120,26 +132,370 @@ type Unknown struct {
+				strMap[k] = str
+			}
+		}
+		return strMap
+	}
+	return nil
+}
+
+func setNestedField(obj map[string]interface{}, value interface{}, fields ...string) {
+	m := obj
+	if len(fields) > 1 {
+		for _, field := range fields[0 : len(fields)-1] {
+			if _, ok := m[field].(map[string]interface{}); !ok {
+				m[field] = make(map[string]interface{})
+			}
+			m = m[field].(map[string]interface{})
+		}
+	}
+	m[fields[len(fields)-1]] = value
+}
+
+func setNestedMap(obj map[string]interface{}, value map[string]string, fields ...string) {
+	m := make(map[string]interface{}, len(value))
+	for k, v := range value {
+		m[k] = v
+	}
+	setNestedField(obj, m, fields...)
+}
+
+func (u *Unstructured) setNestedField(value interface{}, fields ...string) {
+	if u.Object == nil {
+		u.Object = make(map[string]interface{})
+	}
+	setNestedField(u.Object, value, fields...)
+}
+
+func (u *Unstructured) setNestedMap(value map[string]string, fields ...string) {
+	if u.Object == nil {
+		u.Object = make(map[string]interface{})
+	}
+	setNestedMap(u.Object, value, fields...)
+}
+
+// TODO: rewrite this.
+func setField(src reflect.Value, fieldName string, dest interface{}) error {
+	field := src.FieldByName(fieldName)
+	if !field.IsValid() {
+		return fmt.Errorf("%v does not have field %s", src, fieldName)
+	}
+	v, err := conversion.EnforcePtr(dest)
+	if err != nil {
+		return err
+	}
+	if field.Type().AssignableTo(v.Type()) {
+		v.Set(field)
+		return nil
+	}
+	if field.Type().ConvertibleTo(v.Type()) {
+		v.Set(field.Convert(v.Type()))
+		return nil
+	}
+	return fmt.Errorf("couldn't assign/convert %v to %v", field.Type(), v.Type())
+}
+
+func extractOwnerReference(src interface{}) (metatypes.OwnerReference, error) {
+	var ret metatypes.OwnerReference
+	v := src.(map[string]interface{})
+	ret.Kind = getNestedString(v, "kind")
+	ret.Name = getNestedString(v, "name")
+	ret.APIVersion = getNestedString(v, "apiVersion")
+	ret.UID = (types.UID)(getNestedString(v, "uid"))
+	return ret, nil
+}
+
+func setOwnerReference(dest interface{}, src metatypes.OwnerReference) error {
+	v, ok := dest.(map[string]interface{})
+	if !ok {
+		return fmt.Errorf("dest %v has to be of type map[string]interface()", dest)
+	}
+	setNestedField(v, src.Kind, "kind")
+	setNestedField(v, src.Name, "name")
+	setNestedField(v, src.APIVersion, "apiVersion")
+	setNestedField(v, string(src.UID), "uid")
+	return nil
+}
+
+func getOwnerReferences(object map[string]interface{}) (ownerReferences []map[string]interface{}, err error) {
+	field := getNestedField(object, "metadata", "ownerReferences")
+	if field == nil {
+		return ownerReferences, fmt.Errorf("cannot find field metadata.ownerReferences in %v", object)
+	}
+	ownerReferences, ok := field.([]map[string]interface{})
+	if !ok {
+		return ownerReferences, fmt.Errorf("expect metadata.ownerReferences to be a slice in %v", object)
+	}
+	return
+}
+
+func (u *Unstructured) GetOwnerReference(i int) (metatypes.OwnerReference, error) {
```

Comment:
This is a very lengthy approach. I think it would be better to have a `GetOwnerReferences() []metatypes.OwnerReference` and `SetOwnerReferences(refs []metatypes.OwnerReference)`. That way the count, remove, and append go away as it's just slice manipulation at this point.

Also note how none of the other getters return errors. Unfortunate, but I think you should follow the pattern.